### ISSUE TITLE
Reference manual should perhaps explain various basic functional programming concepts

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#18416]
Seems to me it would be better to eventually explain the necessary functional programming concepts in the doc or point to a good description.  This is to make Coq  accessible to a broad audience.  I expect there are only a few concepts needed.  WDYT?

_Originally posted by @jfehrle in https://github.com/coq/coq/pull/18094#discussion_r1429087184_
            
This came up in the context of describing a construct as a fixpoint operator.  Presumably it makes more sense to have a glossary of functional programming concepts, and link there, than to explain fixpoint operator in the section on `rewrite_strat` strategies.

It's not clear to me what concepts need explaining, nor which ones we should explain vs link to, nor what level of background should be assumed, etc.

**Comment by jfehrle** [comment]
I agree this shouldn't be dropped into the middle of the `rewrite_strat` documentation.

> It's not clear to me what concepts need explaining, nor which ones we should explain vs link to, nor what level of background should be assumed, etc.

I couldn't say upfront, either.  The undescribed concepts become apparent over time while reading the doc.

https://github.com/coq/coq/pull/18094#discussion_r1429377735 suggests possible sources for info on `fix`.

**Comment by patrick-nicodemus** [comment]
I would agree with this and I think in particular the section on Ltac2 is not super beginner friendly to people who don't have an understanding of using monads to model programming language semantics. It has taken me a long time to understand the material on that page.

**Comment by patrick-nicodemus** [comment]
I think this is a subtle question and it requires some experience to understand what is meant by an "expression" and a "value." The first link on Google for "expressions" and "values" is here
https://homes.cs.aau.dk/~normark/prog3-03/html/notes/fu-intr-1_themes-expression-section.html

> Expressions 
>    Written by the programmer
>    Will typically involve one or more function calls
>  A Function - as written by the programmer - is itself an expression

> Values
>     Primitive as well as composite
>     A function expression is evaluated to a function object

This definition seems overly simplistic to me. For example, in
```
let x := 3 in
let f := (fun _ => Control.zero Not_found) in
Control.plus (fun () => f x) my_handler
```
When this is evaluated, the let bindings will reduce before Control.plus evaluates its arguments, so we will have `(fun () => (fun _ => Control.zero Not_found) 3)` as the first argument to Control.plus. The user did not manually write  `(fun _ => Control.zero Not_found) 3)`, it was dynamically created during the execution, so it is not an expression, it is a value? But this is not the correct understanding of how Control.plus works. So one needs to understand the concepts of expression and value beyond the level of an elementary extension to use Ltac2.

**Comment by patrick-nicodemus** [comment]
I am not the only one who does not understand this documentation btw.
https://github.com/tchajed/ltac2-tutorial/blob/master/src/ltac2_tutorial.v
> (* The refman explains backtracking as "viewing thunks as lazy lists", and then
the zero and plus operators are the empty list and concatenation operators, and
also these are a "monoid compatible with sequentialization". I don't really
understand this.

I eventually figured this out and I will post my notes on the Zulip soon hopefully and ask for feedback in shaping a PR to the Ltac2 docs.

**Comment by jfehrle** [comment]
I'm glad to help with the wording though I'm not a subject matter expert on
this--though not entirely ignorant either.
